--- ccid-1.4.18/src/ccid_usb.c	2014-09-14 00:27:13.000000000 +0800
+++ ccid-1.4.18.modify/src/ccid_usb.c	2015-04-18 14:36:16.000000000 +0800
@@ -397,10 +397,18 @@
 
 #ifdef USE_COMPOSITE_AS_MULTISLOT
 				static int static_interface = 1;
-
+				static int feitian_static_interface = 0;
+/* 
+	Not sure why the static_interface to let be 1, 
+	but for Feitian R502 dual interface reader, it should be start from 0
+*/
+			
 				/* simulate a composite device as when libudev is used */
+
+/*Add Feitian R502 in*/
 				if ((GEMALTOPROXDU == readerID)
-					|| (GEMALTOPROXSU == readerID))
+					|| (GEMALTOPROXSU == readerID) 
+					|| (FEITIANR502 == readerID))
 				{
 						/*
 						 * We can't talk to the two CCID interfaces
@@ -424,6 +432,11 @@
 
 					/* the CCID interfaces are 1 and 2 */
 					interface_number = static_interface;
+					
+					if (FEITIANR502 == readerID) 
+					{
+						interface_number = feitian_static_interface;
+					}
 				}
 #endif
 				/* is it already opened? */
@@ -566,6 +579,7 @@
 				}
 
 				interface = usb_interface->altsetting->bInterfaceNumber;
+               
 				if (interface_number >= 0 && interface != interface_number)
 				{
 					/* an interface was specified and it is not the
@@ -608,10 +622,14 @@
 #ifdef USE_COMPOSITE_AS_MULTISLOT
 				/* use the next interface for the next "slot" */
 				static_interface++;
-
+				feitian_static_interface++;
+				
 				/* reset for a next reader */
-				if (static_interface > 2)
+				if (static_interface > 2 || feitian_static_interface > 2)
+				{
 					static_interface = 1;
+					feitian_static_interface = 1;
+				}
 #endif
 
 				/* Get Endpoints values*/
@@ -1013,6 +1031,7 @@
 {
 	const struct libusb_interface *usb_interface = NULL;
 	int i;
+ 
 
 	/* if multiple interfaces use the first one with CCID class type */
 	for (i = *num; i < desc->bNumInterfaces; i++)
